libs/providers/in-memory/src/lib/in-memory-provider.ts

```diff
@@ -32,51 +34,70 @@ export class InMemoryProvider implements Provider {
+    return Promise.resolve(this.resolveFlagWithReason<boolean>(flagKey, defaultValue, context, logger));
+  }
 
-    if (typeof flagValue !== 'boolean') {
-      throw new TypeMismatchError();
-    }
+  resolveNumberEvaluation(flagKey: string, defaultValue: number, context?: EvaluationContext, logger?: Logger): Promise<ResolutionDetails<number>> {
+    return Promise.resolve(this.resolveFlagWithReason<number>(flagKey, defaultValue, context, logger));
+  }
 
-    return staticResolution(flagValue);
+  async resolveStringEvaluation(flagKey: string, defaultValue: string, context?: EvaluationContext, logger?: Logger): Promise<ResolutionDetails<string>> {
+    return Promise.resolve(this.resolveFlagWithReason<string>(flagKey, defaultValue, context, logger));
   }
 
-  async resolveStringEvaluation(flagKey: string): Promise<ResolutionDetails<string>> {
-    const flagValue = this.lookupFlagValueOrThrow(flagKey);
+  async resolveObjectEvaluation(flagKey: string, defaultValue: JsonValue, context?: EvaluationContext, logger?: Logger): Promise<ResolutionDetails<any>> {
+    return Promise.resolve(this.resolveFlagWithReason<JsonValue>(flagKey, defaultValue, context, logger));
+  }
 
-    if (typeof flagValue !== 'string') {
-      throw new TypeMismatchError();
+  private resolveFlagWithReason<T extends JsonValue | FlagValueType>(flagKey: string, defaultValue: T, ctx?: EvaluationContext, logger?: unknown): ResolutionDetails<any>{
+    try {
+      const resolutionResult = this.lookupFlagValueOrThrow(flagKey, ctx, logger)
+
+      if (typeof resolutionResult?.value != typeof defaultValue) {
+        throw new TypeMismatchError();
+      }
+
+      return resolutionResult;
+    } catch (error: unknown) {
+      return {
+        value: defaultValue, reason: error instanceof DisabledError ?
+          StandardResolutionReasons.DISABLED
+          : StandardResolutionReasons.DEFAULT
+      }
     }
-
-    return staticResolution(flagValue);
   }
 
-  async resolveNumberEvaluation(flagKey: string): Promise<ResolutionDetails<number>> {
-    const flagValue = this.lookupFlagValueOrThrow(flagKey);
-
-    if (typeof flagValue !== 'number') {
-      throw new TypeMismatchError();
+  private lookupFlagValueOrThrow(flagKey: string, ctx?: EvaluationContext, logger?: unknown): ResolutionDetails<any>{
+    if (!(flagKey in this._flagConfiguration)) {
+      throw new FlagNotFoundError();
     }
+    const flagSpec: Flag = this._flagConfiguration[flagKey];
 
-    return staticResolution(flagValue);
-  }
+    this.throwIfFlagDisabled(flagSpec);
+
+    const isContextEval = ctx && flagSpec?.contextEvaluator;
+    const variant = isContextEval ? flagSpec.contextEvaluator?.(ctx) : flagSpec.defaultVariant;
+
+    const value = variant && flagSpec?.variants[variant];
+
+    if (value === undefined) {
+      throw new TargetingKeyMissingError();
```

Comment:
This is not a `TargetingKeyMissingError`. `TargetingKeyMissingError` is specifically to be used when a provider requires a targeting key but one is not supplied in the context (see https://openfeature.dev/specification/types#error-code).

If the specified variant is missing, this basically means that the flag spec is invalid (the resolved variant doesn't map to a value). I would say this is a `GeneralError` (you could subclass GeneralError and throw a `VariantMissingError` you made up, if you want). You could also make the argument this is a `ParseError`. I'd be fine with either, but `TargetingKeyMissingError` seems wrong.